build_environment.bat

```diff
@@ -23,106 +159,154 @@ if %CONFIG_RELEASE% (
   set RUNTIMELIB="MultiThreadedDebugDLL"
   set _CL_=/MDd
 )
-
+SET BUILDTYPEDIR=%2
 if %DYNAMIC_LIBS% (
   echo Building dynamic libraries - DLLs.
   set STATIC_LIBS=0==1
 ) else (
+  SET BUILDTYPEDIR=%2Static
   echo Building static libraries where there's a choice of projects.
   echo    The rest of the project/solution files need to be set to static manually.
   set STATIC_LIBS=1==1
 )
 
 rem Use all CPU cores
-set CL=/MP
-
-rem Edit to suit your environment
-SET MSVC_PATH=C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC
-SET VCVERSION=140
-SET BUILDTOOLS_PATH=E:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\
-SET BUILDTOOLS_SCRIPT=vcvarsall.bat
-rem If using the Visual C++ Build Tools, uncomment & adjust the following
-rem SET BUILDTOOLS_PATH=C:\Program Files (x86)\Microsoft Visual C++ Build Tools
-rem SET BUILDTOOLS_SCRIPT=vcbuildtools.bat
-rem If building for XP and using the C++ Build Tools package (not full VS,)
-rem   install the Windows 7.1 SDK then uncomment and adjust the following
-rem SET WindowsSdkDir_71A=E:\Program Files\Microsoft SDKs\Windows\v7.1\
+set CL=/MP /FS
 
 SET XCOPY=xcopy /S /Y /I
-SET MSBUILD=msbuild /nologo /m /p:PlatformToolset=v%VCVERSION%_xp /p:RuntimeLibrary=%RUNTIMELIB%
-rem If using the Visual C++ Build Tools, uncomment & adjust the following
-rem SET MSBUILD=msbuild /m /p:PlatformToolset=v140_xp /p:VCTargetsPath="C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\\"
+
+set OLDPATH=%PATH%
 set ROOT_DIR=%CD%
-SET BIN_DIR=%CD%\bin\
-SET LIB_DIR=%CD%\lib\
-SET INCLUDE_DIR=%CD%\include\
+SET BIN_DIR=%ROOT_DIR%\%FOLDER_PLATFORM%\%BUILDTYPEDIR%\bin\
+SET LIB_DIR=%ROOT_DIR%\%FOLDER_PLATFORM%\%BUILDTYPEDIR%\lib\
+SET INCLUDE_DIR=%ROOT_DIR%\%FOLDER_PLATFORM%\%BUILDTYPEDIR%\include\
 SET BUILD_DIR=%CD%\build\
 
-IF "%4" == "nuke" (
-  echo Deleting %LIB_DIR% and %INCLUDE_DIR%
-  rd /S /Q %LIB_DIR%
-  rd /S /Q %INCLUDE_DIR%
-  rem Protoc.exe comes with the repo, so can't delete BIN_DIR
-  rem   (Why don't we build Protoc? -Sean)
-  rem rd /S /Q %BIN_DIR%
-)
-
-set OLDPATH=%PATH%
+REM Everyting prepared. Setup the compiler.
 if %MACHINE_X86% (
-  rem If build OS is 64-bit:
-rem  call "%BUILDTOOLS_PATH%\vcbuildtools.bat" amd64_x86
+IF %OS_IS_64BIT% (
 call "%BUILDTOOLS_PATH%\%BUILDTOOLS_SCRIPT%" amd64_x86
```

Comment:
I have MSVC15 installed on a 64-bit Windows 7 install but it doesn't have this cross-compiler. I had to change this to 'x86' (I also have the problem with @Pegasus-RPG's branch). I think we should probably test for the cross-compiler itself.